engine/src/test/java/org/hibernate/validator/test/internal/engine/cascaded/CustomValueExtractorTest.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.test.internal.engine.cascaded;
+
+import static org.hibernate.validator.testutil.ConstraintViolationAssert.assertCorrectPropertyPaths;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.Valid;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.constraints.Size;
+
+import org.hibernate.validator.HibernateValidator;
+import org.hibernate.validator.constraints.Email;
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+
+/**
+ * @author Gunnar Morling
+ */
+public class CustomValueExtractorTest {
+
+	@Test
+	public void canUseCustomExtractor() throws Exception {
+		Cinema cinema = new Cinema();
+		cinema.visitor = new SomeReference<>( new Visitor() );
+
+		Validator validator = Validation.byProvider( HibernateValidator.class )
+				.configure()
+				.addCascadedValueExtractor( new ReferenceValueExtractor() )
+				.buildValidatorFactory()
+				.getValidator();
+
+		Set<ConstraintViolation<Cinema>> violations = validator.validate( cinema );
+
+		assertCorrectPropertyPaths( violations, "visitor.name" );
+	}
+
+	@Test
+	public void canUseCustomKeyExtractorForMaps() throws Exception {
+		CustomerWithCascadingKeys bob = new CustomerWithCascadingKeys();
+
+		Validator validator = Validation.byProvider( HibernateValidator.class )
+				.configure()
+				.addCascadedValueExtractor( new MapKeyExtractor() )
+				.buildValidatorFactory()
+				.getValidator();
+
+		Set<ConstraintViolation<CustomerWithCascadingKeys>> violations = validator.validate( bob );
+
+		assertCorrectPropertyPaths( violations, "addressByType[key(too short)].value", "addressByType[key(too small)].value" );
```

Comment:
I think I would have preferred something like

```
"addressByType[key('too short')].value"
```

At least for strings and any type that we don't know and will be `toString`ed.
We make exceptions for numbers, booleans, dates